gen/__init__.py

```diff
@@ -48,8 +48,48 @@ def add_roles(cloudconfig, roles):
     return cloudconfig
 
 
-def add_units(cloudconfig, services):
-    cloudconfig['coreos']['units'] += services
+def add_units(cloudconfig, services, cloud_init_implementation='coreos'):
+    '''
+    Takes a services dict in the format of CoreOS cloud-init 'units' and
+    injects into cloudconfig a transformed version appropriate for the
+    cloud_init_implementation.  See:
+    https://coreos.com/os/docs/latest/cloud-config.html for the CoreOS 'units'
+    specification. See: https://cloudinit.readthedocs.io/en/latest/index.html
+    for the Canonical implementation.
+
+    Parameters:
+    * cloudconfig is a dict
+    * services is a list of dict's
+    * cloud_init_implementation is a string: 'coreos' or 'canonical'
+    '''
+    if cloud_init_implementation == 'canonical':
+        for unit in services:
+            unit_name = unit['name']
+            if 'content' in unit:
+                write_files_entry = {'path': '/etc/systemd/system/{}'.format(unit_name),
+                                     'content': unit['content'],
+                                     'permissions': '0644'}
+                cloudconfig.setdefault('write_files', []).append(write_files_entry)
```

Comment:
Why not just always default this right under `cloud_init_implementation == 'canonical'`? Seems like ti would be easier to read (And I would hope an empty write_files would be fine (Not to mention the odds of us ever having no units with content are very slim))